core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -85,6 +95,31 @@ public void deleteFile(String location) {
     io(location).deleteFile(location);
   }
 
+  @Override
+  public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailureException {
+    // peek at the first element to determine the type of FileIO
+    Iterator<String> originalIterator = pathsToDelete.iterator();
+    if (!originalIterator.hasNext()) {
+      return;
+    }
+
+    PeekingIterator<String> iterator = Iterators.peekingIterator(originalIterator);
+    FileIO fileIO = io(iterator.peek());
```

Comment:
why don't we change the function io to return a FileIO which supportsPrefixOperations and BulkOperations if that's a requirement?

Reply:
Thanks for the suggestion, I made this change and removed the typecasts.